py/frequenz/client/weather/_types.py

```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
 
             # Check if the array shape matches the number of filtered times, locations
             # and features
-            if array.shape[0] != len(validity_times_indexes):
+            if validity_times is not None and array.shape[0] != len(validity_times):
                 print(
-                    f"Warning:  The count of validity times in the "
-                    f"array({array.shape[0]}) does not match the expected time "
-                    f"filter count ({validity_times_indexes}."
+                    (
+                        f"Warning:  The count of validity times in the "
+                        f"array({array.shape[0]}) does not match the expected time "
+                        f"filter count ({validity_times_indexes}."
+                    )
                 )
-            if array.shape[1] != len(location_indexes):
+            if locations is not None and array.shape[1] != len(locations):
                 print(
                     f"Warning:  The count of location in the "
                     f"array ({array.shape[1]}) does not match the expected location "
                     f"filter count ({location_indexes})."
                 )
-            if array.shape[2] != len(feature_indexes):
+            if features is not None and array.shape[2] != len(features):
+                print(
+                    f"Warning: The count of features ({array.shape[2]}) does not "
+                    f"match the feature filter count ({feature_indexes})."
+                )
+
+        except Exception as e:
+            raise RuntimeError("Error processing forecast data") from e
+
+        return array
+
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def to_ndarray_vlf_refined(
+        self,
+        validity_times: list[dt.time] | None = None,
+        locations: list[Location] | None = None,
+        features: list[ForecastFeature] | None = None,
+    ) -> np.ndarray[
+        # the shape is known to be 3 dimensional, but the length of each dimension is
+        # not fixed, so we use typing.Any, instead of the usual const generic
+        # parameters.
+        tuple[typing.Any, typing.Any, typing.Any],
+        np.dtype[np.float64],
+    ]:
+        """Convert a Forecast object to numpy array and use NaN to mark irrelevant data.
+
+        If any of the filters are None, all values for that parameter will be returned.
+
+        Args:
+            validity_times: The validity times to filter by.
+            locations: The locations to filter by.
+            features: The features to filter by.
+
+        Returns:
+            Numpy array of shape (num_validity_times, num_locations, num_features)
+
+        Raises:
+            ValueError: If the forecasts data is missing or invalid.
+        """
+        # check for empty forecasts data
+        if not self._forecasts_pb.location_forecasts:
+            raise ValueError("Forecast data is missing or invalid.")
+
+        try:
```

Comment:
As mentioned above I suggest to have only one function, so just a general suggestion: There is a lot of duplicated code here, it would be better to refactor it to share as much common code as possible.